### ISSUE TITLE
Automatically trim trailing whitespaces feature is buggy

### ISSUE THREAD

**Issue by dimitrik-fr** [coteditor/CotEditor#1444]
### Description

Sometimes works properly, sometimes not.

### To Reproduce

1. Go to 'Preferences'
2. Click on 'Edit'
3. Enable 'Automatically trim trailing whitespaces' option

Now: 
1. open any file to edit
2. go to the end of any line and add some spaces to the end
3. press CMD+S to Save the file
4. see that trailing spaces are still here

Another "inconsistency" rather a bug:
1. open any file to edit
2. go to the end of any line and add some spaces to the end
3. move cursor Up to the previous line or Down to the next line
4. press CMD+S to Save the file
5. the trailing spaces on the edited line will disappear
6. but sometimes this "trim" operation will be executed after Save, so your file a second later will be marked as unsaved, and you need to pay attention to this and Save it again to really get it updated on your disk -- which is very confusing..
7. a proper solution would be to be sure to  **always** involve "trim" before Save, if trim feature is enabled, and do Save only once the "trim" operation is really finished

### CotEditor version

4.4.5

### macOS version

Monterey 12.6.3

**Comment by 1024jp** [comment]
First of all, I'd say this is not a bug but by design. See #1124.

6c51f1d1035874821c84fdeb5a823145cfebe813

**Comment by dimitrik-fr** [comment]
Indeed, if this is "by design", then probably makes sense to make it more clear ? -- for ex. add a (?) label with help info near 'Automatically trim trailing whitespaces' option to make users aware about this feature implementation design ?

**Comment by 1024jp** [comment]
Regarding the current auto trailing space trimming feature, it does not state the future relates to the saving.
Therefore, users should not expect the trimming must be done before saving. 

Well...,
 
> First of all -- what it the reason to not trim spaces in the current line ?

Just because it is annoying or confusing for the user when the spaces that have just been input is automatically removed.
In contrast, when the user explicitly removes the trailing spaces by executing the Text > Trim Trailing Whitespaces command, all trailing whitespaces including the current line will be removed.

> Second -- why involve trailing spaces trimming every 3 sec. ?

That's wrong.
CotEditor performs the trimming not *every* 3 seconds but just once 3 seconds later the user stops the text modified (so-called debouncing).
In this way, the performance is not an issue at all.
Syntax highlighting or other procedures by editing have much more weight. 
Moreover, when the trimming is triggered only by saving, it will be seldom done in the environment the Auto Save is disabled.

> - more efficient will be to involve trimming on File Save (explicit Save or Auto Save)

I partly agree with this argument.
Not in the context of efficiency but intuitiveness, such as you expect it.
Furthermore, in fact, the trailing whitespace trimming was once implemented so.
However, at that time, the relationship was removed due to some technical reasons related to usability in AppKit.
Because of those historical reasons, I wanna avoid processing the document content while saving, especially auto-saving.

In conclusion:
1. The current behavior is in spec and works well.
2. Reverting to perform trimming before the saving can be an option. But need to consider it well.

**Comment by 1024jp** [comment]
In the next minor update, the automatic whitespace trimming will be changed to perform additionally every time the user explicitly saves the document.
